README.md
# stomp-param-py!

This is a service for transferring messages using stomp protocol over
ssh connection. It assumes that the accepting part of the ssh
connection links directly into stomp server like this:

    socat STDIO,raw,echo=0 TCP:localhost:61613

## main.py

The main.py script is a basic listener that utilizes the transport. It
connects to the specified server and subscribes to 'brain' queue and
sends messages to 'tg' queue assuming it is a [telegram relay]
implementation.

### Configuration

./main.py script accepts the following command line arguments:


- `-c`, `--config`: configuration file (default /etc/sshstomp/sshstomp.conf)
- `-h`, `--host`: host of stomp server
- `-p`, `--port`: port of stomp server
- `-k`, `--key`: ssh key to use
- `-u`, `--user`: user name
- `--chat-id`: telegram chat_id where messages are sent by default
- `-o`, `--log-file`: file to write log
- `-d`, `--debug`: verbose debug output

Configuration file (if present) is an INI file with a single
`sshstomp` section which could contain the following parameters:

- `host`
- `port`
- `key`
- `user`
- `chat-id`
- `log-file`
- `debug`

These parameters mirror the command line arguments. If both
configuration file entry and command line parameter are present,
command line parameter overrides the configuration file.

`log-file` can be omitted, default behavior is not to write logs to
any file.  `debug` can be omitted, default behavior is to not print
debug messages. All other parameters have to be specified either in a
file or command line.

### Dumb bot implementation

The code currently implements a simple chat bot. All messages are sent
to ChatID specified in configuration unless stated otherwise.

#### Greeting message

The bot sends a greeting message when it is started.

#### Reconnect message

When connection is lost the bot sends a message about it after reconnection.

#### Random ping messages

The bot sends one of random messages with random intervals of about 1 hour.

#### "Don't understand" messages

When bot receives a message from telegram it replies with "Don't
understand" and "What is <quote>?" These messages are sent back to the
chat which was the source of initial message.

### Custom components

The transport uses [modified stomp.py] module which fixes certain
heartbeat/reconnect issues. Testing uses [modified mock ssh server].

[telegram relay]: https://gitlab.com/personal-assistant-bot/infrastructure/pa-tg
[modified stomp.py]: https://github.com/aragaer/stomp.py/tree/heartbeat
[modified mock ssh server]: https://github.com/aragaer/mock-ssh-server/tree/shell


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.0] - 2020-04-05
### Added
- Configuration file
- Logging to file
- Dockerfile
### Changed
- Renamed SshBasedConnection to Connection
- Connection now accepts any instance of Transport
- -c key is now shorthand for --config, not --chat-id
### Fixed
- Send replies to correct chat id, not to owner

## 0.1.0 - 2020-03-26
### Added
- This changelog
- Basic readme
- License
- Initial implementation of ssh transport
- Basic listener

[Unreleased]: https://gitlab.com/personal-assistant-bot/infrastructure/stomp-over-ssh/compare/v0.2.0...master
[0.2.0]: https://gitlab.com/personal-assistant-bot/infrastructure/stomp-over-ssh/compare/v0.1.0...v0.2.0


tests/test_connection.py
import unittest

import stomp

from sshstomp import SshBasedTransport, Connection


class ConnectionTest(unittest.TestCase):

    def test_create(self):
        tr = SshBasedTransport("localhost",
                               12345,
                               "username",
                               "files/test.key")
        conn = Connection(tr,
                          heartbeats=(120000, 180000),
                          auto_content_length=True,
                          heart_beat_receive_scale=1.5)

        self.assertIsInstance(conn, stomp.Connection12)

    def test_create_defaults(self):
        tr = SshBasedTransport("localhost",
                               12345,
                               "username",
                               "files/test.key")
        Connection(tr)


tests/test_ssh_transport.py
import unittest

import mockssh
import stomp

from sshstomp import SshBasedTransport


_USERS = {'username': 'files/test.key'}


class SshTransportTest(unittest.TestCase):

    def setUp(self):
        self.serv = mockssh.Server(_USERS)
        self.serv.__enter__()
        self.addCleanup(self.serv.__exit__)
        self.tr = SshBasedTransport("localhost", self.serv.port,
                                    "username", 'files/test.key')

    def test_create(self):
        self.assertIsInstance(self.tr, stomp.transport.BaseTransport)
        self.assertFalse(self.tr.is_connected())

    def test_vhost(self):
        # stomp.py requires this parameter
        self.assertEqual(self.tr.vhost, f"username@localhost:{self.serv.port}")

    def test_send_not_connected(self):
        with self.assertRaises(stomp.exception.NotConnectedException):
            self.tr.send(b'echo hello, world')

    def test_connect(self):
        self.tr.attempt_connection()

        self.assertTrue(self.tr.is_connected())

        # stomp.py requires this parameter to be set after connection
        self.assertEqual(self.tr.current_host_and_port, ("localhost", self.serv.port))

    def test_send(self):
        self.tr.attempt_connection()

        data = b'echo hello, world'
        self.tr.send(data)

    def test_recv(self):
        self.tr.attempt_connection()
        self.tr.send(b'hello, world')

        result = self.tr.receive()

        self.assertEqual(result, b'hello, world')

    def test_cleanup(self):
        # only closes shell, doesn't disconnect the socket
        self.tr.attempt_connection()
        self.tr.send(b'hello, world')

        self.tr.cleanup()

        self.assertFalse(self.tr.is_connected())

        self.tr.attempt_connection()
        self.tr.send(b'yo')
        self.assertEqual(self.tr.receive(), b'yo')

    def test_disconnect(self):
        # disconnects the socket completely
        self.tr.attempt_connection()
        self.tr.send(b'hello, world')

        self.tr.disconnect_socket()

        self.assertFalse(self.tr.is_connected())

        self.tr.attempt_connection()
        self.tr.send(b'yo')
        self.assertEqual(self.tr.receive(), b'yo')


tests/test_ssh_connection.py
import unittest

import stomp

from sshstomp import SshBasedConnection


class SshConnectionTest(unittest.TestCase):

    def test_create(self):
        conn = SshBasedConnection("localhost",
                                  12345,
                                  "username",
                                  "files/test.key",
                                  heartbeats=(120000, 180000),
                                  auto_content_length=True,
                                  heart_beat_receive_scale=1.5)

        self.assertIsInstance(conn, stomp.Connection12)

    def test_create_defaults(self):
        SshBasedConnection("localhost",
                           12345,
                           "username",
                           "files/test.key")


heartbeat.py
#!/usr/bin/env python3
import logging
import socket
import sys
import time

import stomp


_LOGGER = logging.getLogger(__name__)


class UnixSocketTransport(stomp.transport.BaseTransport):

    def __init__(self, sockname):
        super().__init__()
        self.socket = None
        self.current_host_and_port = sockname
        self.vhost = f'localhost'
        self.connected = False

    def is_connected(self):
        return self.connected

    def disconnect_socket(self):
        _LOGGER.info("disconnect socket is called")
        self.socket.close()

    def send(self, encoded_frame):
        if not self.connected:
            raise stomp.exception.NotConnectedException()
        return self.socket.sendall(encoded_frame)

    def receive(self):
        result = self.socket.recv(1024)
        _LOGGER.debug("Got %s", result)
        return result

    def cleanup(self):
        _LOGGER.info("cleanup called")
        self.socket.shutdown(socket.SHUT_RDWR)
        self.socket.detach()
        self.connected = False

    def attempt_connection(self):
        if self.socket is not None:
            _LOGGER.info("closing old socket")
            self.socket.close()
        self.socket = socket.socket(socket.AF_UNIX)
        _LOGGER.info("connecting")
        self.socket.connect(self.current_host_and_port)
        self.connected = True

    def set_ssl(self,
                for_hosts=[],
                key_file=None,
                cert_file=None,
                ca_certs=None,
                cert_validator=None,
                ssl_version=None,
                password=None):
        pass

    def get_ssl(self, host_and_port=None):
        return None


class UnixSocketConnection(stomp.connect.StompConnection12):
    def __init__(self,
                 sockname,
                 heartbeats=(0, 0),
                 auto_content_length=True,
                 heart_beat_receive_scale=1.5):
        transport = UnixSocketTransport(sockname)
        stomp.connect.BaseConnection.__init__(self, transport)
        stomp.protocol.Protocol12.__init__(self, transport, heartbeats, auto_content_length,
                                           heart_beat_receive_scale=heart_beat_receive_scale)


class MyListener(stomp.ConnectionListener):

    def __init__(self, conn, pub, sub):
        super().__init__()
        self.conn = conn
        self.done = False
        self.pub = pub
        self.sub = sub
        self.queue = []

    def on_error(self, headers, message):
        _LOGGER.info('received an error %s', message)

    def on_connected(self, headers, message):
        _LOGGER.info('connected')
        self.conn.subscribe(self.sub, id=1)
        while self.queue:
            item = self.queue.pop(0)
            try:
                self.conn.send(body=item, destination=self.pub)
            except:
                self.queue.insert(0, item)

    def on_message(self, headers, message):
        _LOGGER.info('received a message %s', message)

    def on_disconnected(self):
        _LOGGER.info('disconnected')
        if not self.done:
            _LOGGER.info('sleep 5')
            time.sleep(5)
            self.queue.append("reconnect")
            self.conn.connect()


def main():
    conn = UnixSocketConnection("/tmp/xtomp.sock", heartbeats=(120000,75000))
    listener = MyListener(conn, 'memq', 'memq')
    listener.queue.append('ping')
    conn.set_listener('', listener)
    conn.connect()
    while True:
        try:
            time.sleep(240)
#            conn.send(body='ping', destination='memq')
        except KeyboardInterrupt:
            break
    listener.done = True
    conn.disconnect()


if __name__ == '__main__':
    logging.basicConfig(stream=sys.stderr,
                        level=logging.DEBUG,
                        format="%(asctime)-8s:%(levelname)5s:%(name)s: %(message)s",
                        datefmt="%H:%M:%S")
    handler = logging.FileHandler('heartbeat.log')
    handler.setFormatter(logging.Formatter("%(asctime)-8s:%(levelname)5s:%(name)s: %(message)s"))
    logging.getLogger('').addHandler(handler)
    main()


sshstomp/__init__.py
import logging
import time

import paramiko
import stomp


_LOGGER = logging.getLogger(__name__)


class SshBasedTransport(stomp.transport.BaseTransport):
    """
    STOMP transport that utilizes ssh connection instead of socket.
    """

    def __init__(self, hostname, port, username, key):
        super().__init__()
        self.client = paramiko.SSHClient()
        self.client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
        self.hostname = hostname
        self.port = port
        self.username = username
        self.vhost = f"{username}@{hostname}:{port}"
        self.channel = None
        self.key = key

    def is_connected(self):
        return self.channel is not None and not self.channel.closed

    def attempt_connection(self):
        sleep_time = 8
        while not self.is_connected():
            try:
                transport = self.client.get_transport()
                if transport is None or not transport.is_active():
                    self.client.connect(self.hostname, self.port, self.username, key_filename=self.key)
                    self.current_host_and_port = (self.hostname, self.port)

                _LOGGER.debug("transport connected")
                if self.channel is None or self.channel.closed:
                    self.client.get_transport().set_keepalive(60)
                    self.channel = self.client.invoke_shell()
                    time.sleep(0.2)
            except paramiko.ssh_exception.SSHException as ex:
                time.sleep(sleep_time)
                if sleep_time < 100:
                    sleep_time *= 2

    def send(self, encoded_frame):
        if not self.is_connected():
            raise stomp.exception.NotConnectedException()
        self.channel.sendall(encoded_frame)

    def receive(self):
        result = self.channel.recv(4096)
        _LOGGER.debug("Got %s", result)
        return result

    def cleanup(self):
        self.channel.close()
        self.channel = None

    def disconnect_socket(self):
        self.channel.close()
        self.client.close()


class Connection(stomp.Connection12):
    """
    STOMP connection that uses custom transport
    """

    def __init__(self, transport,
                 heartbeats=(0,0),
                 auto_content_length=True,
                 heart_beat_receive_scale=1.5):
        stomp.connect.BaseConnection.__init__(self, transport)
        stomp.protocol.Protocol12.__init__(self, transport, heartbeats, auto_content_length,
                                           heart_beat_receive_scale=heart_beat_receive_scale)


main.py
#!/usr/bin/env python3
import argparse
import configparser
import logging
import json
import os
import sys
import random
import time

import paramiko
import stomp

from sshstomp import Connection, SshBasedTransport


_LOGGER = logging.getLogger(__name__)


class KeepAlivesFilter:
    def filter(self, record):
        return 'keepalive@' not in record.msg


def setup_logging(level, logfile):
    paramiko.util.get_logger('paramiko.transport').addFilter(KeepAlivesFilter())
    log_format = "%(asctime)-8s:%(levelname)5s:%(name)s: %(message)s"
    logging.basicConfig(stream=sys.stderr,
                        level=level,
                        format=log_format,
                        datefmt="%H:%M:%S")
    if logfile is not None:
        handler = logging.FileHandler(os.path.expanduser(logfile))
        handler.setFormatter(logging.Formatter(log_format))
        logging.getLogger('').addHandler(handler)


class MyListener(stomp.ConnectionListener):

    def __init__(self, conn, chat_id):
        self.conn = conn
        self.done = False
        self.chat_id = chat_id

    def on_error(self, headers, message):
        print('received an error', message)

    def on_connected(self, headers, message):
        print('connected')
        self.conn.subscribe('brain', id=1)

    def on_message(self, headers, message):
        print('received a message', message)
        msg = json.loads(message)
        if msg['from']['channel'] == 'tg':
            sender = f"tg:{msg['from']['user_id']}"
        else:
            sender = msg['from']['channel']
        print(f"Message from {sender}:", msg['text'])

        if msg['from']['channel'] == 'tg':
            if msg['from']['user_id'] == msg['from']['chat_id']: # private message
                self.conn.send(body=make_message(["Я не поняла",
                                                  f"А что значит \"{msg['text']}\"?"],
                                                 chat_id=msg['from']['chat_id']),
                               destination='tg')

    def on_disconnected(self):
        print('disconnected')
        if not self.done:
            _LOGGER.info('sleep 5')
            time.sleep(5)
            self.conn.connect()
            self.conn.send(body=make_message(random.choice(["Ой, что-то поломалось",
                                                            "Тут что-то сломалось",
                                                            "Кажется, связь моргнула"]),
                                             chat_id=self.chat_id),
                           destination='tg')


def make_message(text, chat_id):
    return json.dumps({"from": {"channel": "brain", "name": "niege"},
                       "to": {"channel": "tg", "chat_id": chat_id},
                       "text": text})


def coalesce(x1, x2):
    return x2 if x1 is None else x1


def get_config(config_file_arg):
    config_file = os.path.expanduser(config_file_arg)

    config = configparser.ConfigParser()
    if os.path.exists(config_file):
        config.read(config_file)
    else:
        config.read_dict({'sshstomp':{}})

    return config['sshstomp']


def process_config(args):
    config = get_config(args.config)

    log_file = coalesce(args.log_file, config.get('log-file', fallback=None))
    if args.debug or config.getboolean('debug', fallback=False):
        log_level = logging.DEBUG
    else:
        log_level = logging.INFO

    setup_logging(log_level, log_file)

    host = coalesce(args.host, config.get('host'))
    port = coalesce(args.port, config.getint('port'))
    user = coalesce(args.user, config.get('user'))
    keyfile = os.path.expanduser(coalesce(args.key, config.get('key')))
    chat_id = coalesce(args.chat_id, config.getint('chat-id'))
    return host, port, user, keyfile, chat_id


def main(args):
    host, port, user, keyfile, owner = process_config(args)
    transport = SshBasedTransport(host, port, user, keyfile)
    conn = Connection(transport, heartbeats=(120000,180000))
    listener = MyListener(conn, owner)
    conn.set_listener('', listener)
    conn.connect()
    conn.send(body=make_message("я тут", chat_id=owner), destination='tg')
    while True:
        try:
            time.sleep(3600+random.randint(-600, 600))
            text = random.choice(["Привет",
                                  "Ой, приветик",
                                  "Чем занят?",
                                  "Мне скучно"])
            conn.send(body=make_message(text, chat_id=owner),
                      destination='tg')
        except KeyboardInterrupt:
            break
    listener.done = True
    conn.disconnect()

if __name__ == '__main__':
    parser = argparse.ArgumentParser(add_help=False)
    parser.add_argument('-h', '--host')
    parser.add_argument('-p', '--port', type=int)
    parser.add_argument('-k', '--key')
    parser.add_argument('-u', '--user')
    parser.add_argument('--chat-id', type=int)
    parser.add_argument('-c', '--config', default="/etc/sshstomp/sshstomp.conf")
    parser.add_argument('-d', '--debug', action='store_true')
    parser.add_argument('-o', '--log-file')

    args = parser.parse_args()

    main(args)
